File: test/scripts/08-closures-allocations/incorrect.py
#!/usr/bin/env python3
"""
INCORRECT: Closures in hot path - capture creates heap allocations.
Resume: "Avoid Closures in Hot Paths"
"""
import os
import signal
import time
N = 1_000_000
DURATION_SEC = int(os.environ.get("DURATION_SEC", "300"))
stop = False

def main():
    global stop
    threshold = 50
    data = list(range(100))
    print("INCORRECT: listcomp with closure (captures threshold)")
    signal.signal(signal.SIGINT, lambda s, f: globals().__setitem__('stop', True))
    signal.signal(signal.SIGTERM, lambda s, f: globals().__setitem__('stop', True))
    start = time.perf_counter()
    laps = 0
    while not stop and (time.perf_counter() - start) < DURATION_SEC:
        total = 0
        for _ in range(N):
            filtered = [x for x in data if x > threshold]
            total += sum(filtered)
        laps += 1
    print(f"Stopped: {laps} laps")

if __name__ == "__main__":
    main()

File: test/scripts/08-closures-allocations/incorrect.js
const N = 1_000_000;
const DURATION_SEC = parseInt(process.env.DURATION_SEC || '300', 10);
const data = Array.from({ length: 100 }, (_, i) => i);
let laps = 0;
const start = Date.now();
const threshold = 50;
console.log('INCORRECT: filter with closure');
const id = setInterval(() => {
  if (Date.now() - start >= DURATION_SEC * 1000) {
    clearInterval(id);
    console.log('Stopped:', laps, 'laps');
    return;
  }
  let total = 0;
  for (let i = 0; i < N; i++) {
    const filtered = data.filter(x => x > threshold);
    total += filtered.reduce((a, b) => a + b, 0);
  }
  laps++;
}, 0);
process.on('SIGINT', () => { clearInterval(id); process.exit(0); });
